plugins/chat/app/jobs/regular/auto_join_channel_batch.rb

```diff
@@ -56,26 +58,36 @@ def create_memberships_query(category)
           uccm.chat_channel_id = :chat_channel_id AND uccm.user_id = users.id
       SQL
 
-      query += <<~SQL if category.read_restricted?
+      query += <<~SQL if restricted
           INNER JOIN group_users gu ON gu.user_id = users.id
-          LEFT OUTER JOIN category_groups cg ON cg.group_id = gu.group_id
+          LEFT OUTER JOIN category_groups cg ON cg.group_id = gu.group_id AND
+          cg.permission_type <= :permission_type
         SQL
 
       query += <<~SQL
         WHERE (users.id >= :start AND users.id <= :end) AND
           users.staged IS FALSE AND users.active AND
           NOT EXISTS(SELECT 1 FROM anonymous_users a WHERE a.user_id = users.id) AND
           (suspended_till IS NULL OR suspended_till <= :suspended_until) AND
-          (last_seen_at > :last_seen_at) AND
+          (last_seen_at IS NULL OR last_seen_at > :last_seen_at) AND
           uo.chat_enabled AND
           uccm.id IS NULL
       SQL
 
-      query += <<~SQL if category.read_restricted?
-          AND cg.category_id = :channel_category
+      query += <<~SQL if restricted
+      AND cg.category_id = :channel_category
         SQL
 
       query += "RETURNING user_chat_channel_memberships.user_id"
     end
+
+    def join_restricted?(category)
+      category.read_restricted? ||
+        CategoryGroup.exists?(
```

Comment:
This doesn't seem quite right, since a category can have multiple groups, each of which could potentially be `readonly`, and `everyone` doesn't even need to be here either:

![image](https://user-images.githubusercontent.com/920448/220521457-6de6ce50-187a-410d-98b3-a1cd74ca64a4.png)

I _think_ rather than even having this `join_restricted` you just unconditionally join on `category_groups` and always specify `cg.category_id = :channel_category`, since the user must be in 1 or more groups which have `create_post` or `full` category group permissions for the category to be able to join/auto-join the channel.

Reply:
If there are no `category_groups` records then the category is only visible to staff...in that case you can probably conditionally remove those category group parts of the query and add in `user.admin OR user.moderator` clauses.

Reply:
I agree. I still had to make some exceptions for two reasons:
1. When `everyone` has `full` access, there is no `CategoryGroup` for that relationship. The good thing is when there's no group set through the UI, it creates a `CategoryGroup` for `staff`
2. The user membership to `everyone` is implicit. There is no `GroupUser` for `everyone`

Reply:
Yes thank you, I made these initial comments before I had to discover all this on my own with my auto removal stuff :facepalm: